contracts/vaults/dstake/sdUSDBalanceChecker.sol

```diff
@@ -0,0 +1,226 @@
+ *      unlike dLEND tokens since sdUSD represents direct vault shares.
+ */
+contract sdUSDBalanceChecker is IBalanceChecker, AccessControl {
+    /// @notice The sdUSD token address (ERC4626 vault) on Fraxtal mainnet
+    address public constant SD_USD_TOKEN = 0x58AcC2600835211Dcb5847c5Fa422791Fd492409;
+
+    /// @notice Mapping from external token to its corresponding sdUSD token
+    mapping(address => address) public externalSourceToSdUSDToken;
+
+    /**
+     * @param initialAdmin The address that will be granted the DEFAULT_ADMIN_ROLE
+     */
+    constructor(address initialAdmin) {
+        require(initialAdmin != address(0), "INVALID_ADMIN_ADDRESS");
+        _grantRole(DEFAULT_ADMIN_ROLE, initialAdmin);
+        
+        // Map the sdUSD token to itself for direct queries
+        externalSourceToSdUSDToken[SD_USD_TOKEN] = SD_USD_TOKEN;
+    }
+
+    /**
+     * @notice Maps an external token to its corresponding sdUSD token
+     * @param externalToken The address of the external token
+     * @param sdUSDToken The address of the corresponding sdUSD token (should be ERC4626 compliant)
+     */
+    function mapExternalSource(
+        address externalToken,
+        address sdUSDToken
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(sdUSDToken != address(0), "INVALID_SDUSD_TOKEN_ADDRESS");
+        externalSourceToSdUSDToken[externalToken] = sdUSDToken;
+    }
+
+    /**
+     * @dev Internal helper function to validate token and get necessary details
+     * @param token The token address to validate
+     * @return validToken The validated token address (sdUSD token)
+     * @return originalToken The original token address
+     * @return isExternalToken Whether the token is external
+     * @return underlyingAsset The underlying asset address from the ERC4626 vault
+     */
+    function _validateTokenAndGetDetails(
+        address token
+    )
+        internal
+        view
+        returns (
+            address validToken,
+            address originalToken,
+            bool isExternalToken,
+            address underlyingAsset
+        )
+    {
+        originalToken = token;
+        address mappedSdUSDToken = externalSourceToSdUSDToken[token];
+        isExternalToken = mappedSdUSDToken != address(0);
+        validToken = isExternalToken ? mappedSdUSDToken : token;
+
+        try IERC4626(validToken).asset() returns (address asset) {
+            underlyingAsset = asset;
+        } catch {
+            if (isExternalToken) {
+                revert InvalidERC4626Token(mappedSdUSDToken);
+            } else {
+                revert ExternalTokenNotMapped(originalToken);
+            }
+        }
+    }
+
+    /**
+     * @inheritdoc IBalanceChecker
+     */
+    function tokenBalances(
+        address token,
+        address[] memory addresses
+    ) external view override returns (uint256[] memory result) {
+        require(addresses.length <= 1000, "TOO_MANY_ADDRESSES");
```

Comment:
We shouldn't have this require, we only need to support at least 1000 addresses within the gas limit, which this should comfortably do already, not a maximum